gateway/src/apicast/policy/rate_limiting_to_service/rate_limiting_to_service.lua

```diff
@@ -0,0 +1,201 @@
+local policy = require('apicast.policy')
+local _M = policy.new('Rate Limiting to Service Policy')
+
+local ngx_semaphore = require "ngx.semaphore"
+local limit_traffic = require "resty.limit.traffic"
+local resty_redis = require('resty.redis')
+local tonumber = tonumber
+local next = next
+local shdict_key = 'limitter'
+
+local new = _M.new
+
+function _M.new(config)
+  local self = new()
+  self.config = config or {}
+  self.limitters = config.limitters
+  self.redis_info = config.redis_info
+  return self
+end
+
+local function redis_shdict(host, port, db)
+  local redis = assert(resty_redis:new())
+
+  local ok, connerr = redis:connect(host or '127.0.0.1', port or 6379)
+  if not ok then
+    return nil, connerr
+  end
+
+  ok = redis:select(db or 0)
+  if not ok then
+    return nil, "failed to select db"
+  end
+
+  return {
+    incr = function(_, key, value, init)
+      if not init then
+        return redis:incrby(key, value), nil
+      end
+      redis:setnx(key, init)
+      return redis:incrby(key, value), nil
+    end,
+    set = function(_, key, value)
+      return redis:set(key, value)
+    end,
+    expire = function(_, key, exptime)
+      local ret = redis:expire(key, exptime)
+      if ret == 0 then
+        return nil, "not found"
+      end
+      return true, nil
+    end,
+    get = function(_, key)
+      local val = redis:get(key)
+      if type(val) == "userdata" then
+        return nil
+      end
+      return val
+    end
+  }
+end
+
+local function try(f, catch_f)
+  local status, exception = pcall(f)
+  if not status then
+    catch_f(exception)
+  end
+end
+
+function _M:access()
+  local limitters = {}
+  local keys = {}
+  local states = {}
+
+  local limitters_limit_conn = {}
+  local keys_limit_conn = {}
+
+  local limitters_limit_conn_committed = {}
+  local keys_limit_conn_committed = {}
+
+  if not self.redis_info then
+    ngx.log(ngx.ERR, "No Redis information.")
+    return ngx.exit(500)
```

Comment:
I think we want this policy to work without redis. It should just work with shared memory. Redis should be opt-in.